sw/host/opentitanlib/src/proxy/socket_server.rs

```diff
@@ -0,0 +1,298 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use mio::event::Event;
+use mio::net::TcpListener;
+use mio::net::TcpStream;
+use mio::{Events, Interest, Poll, Token};
+use mio_signals::{Signal, SignalSet, Signals};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::HashMap;
+use std::io::{ErrorKind, Read, Write};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use super::CommandHandler;
+
+const BUFFER_SIZE: usize = 8192;
+const EOL_CODE: u8 = b'\n';
+
+fn get_next_token() -> Token {
+    static TOCKEN_COUNTER: AtomicUsize = AtomicUsize::new(0);
+    Token(TOCKEN_COUNTER.fetch_add(1, Ordering::Relaxed))
+}
+
+/// This struct listens on a TCP socket, and maintains a number of concurrent connections,
+/// receiving serialized JSON representations of `Msg`, passing them to the given
+/// `CommandHandler` to obtain responses to be sent as socket flow contol permits.  Note that
+/// this implementaion is not specific to (and does not refer to) any particular protocol.
+pub struct JsonSocketServer<Msg: DeserializeOwned + Serialize, T: CommandHandler<Msg>> {
+    command_handler: T,
+    poll: Poll,
+    socket: TcpListener,
+    socket_token: Token,
+    signals: Signals,
+    signal_token: Token,
+    connection_map: HashMap<Token, Connection>,
+    exit_requested: bool,
+    phantom: std::marker::PhantomData<Msg>,
+}
+
+impl<Msg: DeserializeOwned + Serialize, T: CommandHandler<Msg>> JsonSocketServer<Msg, T> {
+    pub fn new(command_handler: T, port: u16) -> Result<Self> {
+        let poll = Poll::new()?;
+        let socket_token = get_next_token();
+        let addr = std::net::SocketAddr::new(
+            std::net::IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0)),
```

Comment:
I'd go and `use` all these net names, tbh. You can even write
```
SocketAddr::from(([0; 4], port))
```
since IpAddr can be built from a [u8; 4] and SocketAddr can be built from a pair of that and u16, via From.

Reply:
Done